corehq/util/zip_utils.py

```diff
@@ -66,3 +73,74 @@ def get(self, request, *args, **kwargs):
             response['Content-Length'] = os.path.getsize(fpath)
             set_file_download(response, self.zip_name)
             return response
+
+
+def iter_files_async(include_multimedia_files, include_index_files, app):
+    file_iterator = lambda: iter([])
+    errors = []
+    if include_multimedia_files:
+        from corehq.apps.hqmedia.views import iter_media_files
+        app.remove_unused_mappings()
+        file_iterator, errors = iter_media_files(app.get_media_objects())
+    if include_index_files:
+        from corehq.apps.app_manager.views import iter_index_files
+        if app.is_remote_app():
+            file_iterator, errors = iter_index_files(app)
+        else:
+            index_files, index_file_errors = iter_index_files(app)
+            if index_file_errors:
+                errors.append(index_file_errors)
+            file_iterator = itertools.chain(file_iterator, index_files)
+
+    return file_iterator, errors
+
+
+def make_zip_tempfile_async(include_multimedia_files,
+                            include_index_files, app, download_id, compress_zip=False, filename="commcare.zip"):
+    errors = []
+    compression = zipfile.ZIP_DEFLATED if compress_zip else zipfile.ZIP_STORED
+
+    use_transfer = False
+    if transfer_enabled() and os.path.isdir(settings.TRANSFER_FILE_DIR):
+        fpath = os.path.join(settings.TRANSFER_FILE_DIR, "{}{}".format(app._id, app.version))
+        use_transfer = True
+    else:
+        _, fpath = tempfile.mkstemp()
+
+    if not os.path.isfile(fpath):  # Don't rebuild the file if it is already there
+        files, errors = iter_files_async(include_multimedia_files, include_index_files, app)
+        with open(fpath, 'wb') as tmp:
+            with zipfile.ZipFile(tmp, "w") as z:
+                for path, data in files:
+                    # don't compress multimedia files
+                    extension = os.path.splitext(path)[1]
+                    file_compression = zipfile.ZIP_STORED if extension in MULTIMEDIA_EXTENSIONS else compression
+                    z.writestr(path, data, file_compression)
+
+    return expose_file_download(fpath,
+                                mimetype='application/zip' if compress_zip else 'application/x-zip-compressed',
+                                download_id=download_id,
+                                content_disposition='attachment; filename="{fname}"'.format(fname=filename),
+                                use_transfer=use_transfer), errors
+
+
+class DownloadZipAsync(DownloadZip):
+    include_multimedia_files = False
+    include_index_files = False
+
+    def get(self, request, *args, **kwargs):
```

Comment:
Should add an assert here that either `include_multimedia_files` or `include_index_files` is true